audacity/audacity
# Rewrite track type switch

Resolves: *(direct link to the issue)*

Use TypeList, Tuple, and Callable to clean up some old metaprogramming implementing Track::TypeSwitch.

It's not simple -- there are three levels of TypeList folds now -- but the intent of the recursions is more clearly expressed that way.

A previous mutual recursion (!) of templates goes away.

Type switch case functions accept references, not pointers. Some std::function overhead can be eliminated for type switches that use fallthroughs, after their call sequence changes too.  Optional extra arguments can be passed (not needed for existing uses that visit tracks).

Still to do: add unit tests.

I expect to re-use this in other improvements to the Registry system, to make visitation code simpler and type-checked.

QA: Make a project with various track types (audio, label, MIDI, time) and briefly exercise these things, where possible selecting tracks of all different kinds:
- [x] A per-track effect or generator, and verify sync lock adjustment of unselected tracks -- see below
- [x] Project persistency
- [x] Making stereo track from monos
- [x] Pasting label and MIDI data
- [x] MIDI playback
- [x] Import audio
- [x] "Get Info" macro command (with all choices)
- [x] "Set Clip"
- [x] "Set Envelope"
- [x] Effects and generators (and verify sync lock for generators):  AutoDuck, ChangeSpeed, Silence, Repeat, Reverse, ChangeTempo (both checkbox settings), Truncate Silence, any Nyquist effect
- [x] Paste in the Edit menu
- [x] Cut, Split Cut, Split Delete, Trim, Split New in Edit menu
- [x] In Label sub-menus of Edit: cut, split cut, split delete, silence, copy, split, join, disjoin
- [x] Track and Extra > Track menus:  sort, pan, pan left, pan right, gain, gain up, gain down, mute, solo
- [x] Envelope editing (both audio and time tracks)
- [x] Dragging time selection and box spectral selection, double click clip selection

More explanation of steps for the first checkbox:
- Generate some sound
- Turn on sync lock, and make other audio and label tracks below
- Select a portion of the first track only, and Gererate > Tone or Chirp
- Change duration from the default (the selection length), either longer or shorter
- OK, and observe correct generation and sync lock adjustment




Recommended:
